### ISSUE TITLE
How to set additionalProperties=False?

### ISSUE THREAD

**Issue by sbrandtb** [toumorokoshi/jsonschema-extractor#3]
I want to set [`additionalProperties = False`](https://json-schema.org/understanding-json-schema/reference/object.html#properties) on the generated jsonschema objects.

Before I create a merge request, what do you think would be the the best approach to this?

**Comment by toumorokoshi** [comment]
Hey @sbrandtb! I think it's fair to want that customization: I suppose whether that schema is valid or not is entirely dependent on the use case.

Regarding how to do this work: it looks like the existing behavior is to have that field blank. To make it possible to configure this for all extractors, I'd suggest putting this SchemaExtractorSet's extract method:

https://github.com/toumorokoshi/jsonschema-extractor/blob/master/jsonschema_extractor/extractor_set.py#L19

And allowing a kwarg in the constructor like "additional_properties=None" (none implies leave it blank).

It wouldn't let you modify the default extractor, but it does allow for providing a custom one.

Thoughts on the approach?

**Comment by sbrandtb** [comment]
@toumorokoshi Hm, that would require us passing around this parameter and potentially new ones all over the place (i.e. in all calls of `extract` and that are quite a lot) and doesn't feel really right.

I'm wondering if we could get rid of the instantiation of the extractors [here](https://github.com/toumorokoshi/jsonschema-extractor/blob/master/jsonschema_extractor/__init__.py#L5) and [here](https://github.com/toumorokoshi/jsonschema-extractor/blob/master/jsonschema_extractor/__init__.py#L9), because the constructors take no arguments anyways and instead create the extractor instance in the extractor set.

**Comment by toumorokoshi** [comment]
I think instantiation in the constructor would be fine, if there is no list
passed in already.

I don't think you'll have to pass the parameter around, as that field can
be generically applied to any jsonschema. Without modifying the extractors,
you could add additional processing to the ExtractorSet's extract method
like:


```
result = self[typ].extract(self, typ)
if self._additional_properties:
    result["additionalProperties"] = self._additional_properties
return result
```

There's a strong argument to add global processing and configuration of the
jsonschema in the extractorset itself.

**Comment by sbrandtb** [comment]
Maybe I understood you wrong, but `additionalProperties` is not a global configuration, but a property of each and every object type and it does not cascade to children. See example:

```json
{
  "type": "object",
  "properties": {
    "sub": {
      "type": "object",
      "properties": {
        "name": {
          "type": "string"
        }
      },
      "required": [
        "name"
      ]
    }
  },
  "required": [
    "sub"
  ],
  "additionalProperties": false
}
```

This validates perfectly:

```json
{
  "sub": {
    "name": "foo",
    "i should not be here": "but i am"
  }
}
```

But not anymore if you add `additionalProperties: false` to the sub object:

```json
{
  "type": "object",
  "properties": {
    "sub": {
      "type": "object",
      "properties": {
        "name": {
          "type": "string"
        }
      },
      "required": [
        "name"
      ],
      "additionalProperties": false
    }
  },
  "required": [
    "sub"
  ],
  "additionalProperties": false
}
```

**Comment by sbrandtb** [comment]
Another question would be: Should `additionalProperties` be inherent to the class/object extracted or global? For instance, `required` is deducted implicitly from the default value of a field. So, should we deduct the `additionalProperties` from the class? The problem I see here is that it would be very cumbersome to put that property in *every* class.

**Comment by toumorokoshi** [comment]
Thanks for the detailed description. Yes, I see the solution I provided may not be granular enough.

In the case where this is a property to set for all objects (including nested), I think my solution would work. If you look at the code for the extractors themselves, they call the extractor_set.extract method when performing schema extraction of nested objects.

As such, even nested attrs objects would return back "additionalProperties" with the appropriate setting.

All that said, I think this does become a trickier problem because of the need to better understand if the property is needed per-class.

in attrs extractor, usage of the extractor passed in (should be extractor_set):

https://github.com/toumorokoshi/jsonschema-extractor/blob/master/jsonschema_extractor/attrs_extractor.py#L44

Regarding what it would look like if that additionalProperties was more granular: I think it depends on the serializer. In some cases (like schematics) the choice of whether to accept unknown keys can be specified in the object. Outside of that I think a global default as determined by the extractor_set generally suffices. If we really need to expose more class-specific instances we could use decorators or configure the extractor_set to treat very specific classes differently.

Globally effectively means making it a parameter on the extractor_set, since the extract methods just defer to an extractor_set object. I'm not a fan of a parameter that is global beyond that: in my experience making completely global switches (i.e. not using an instanced version as a default and using that as a global) results in difficulty when one needs to have the global switch on for some situations, but not for others.

**Comment by sbrandtb** [comment]
I started using [pydantic](https://pydantic-docs.helpmanual.io/) for my task because I realized that it is very much made for my use case (serialize/deserialize/validate) and offers way more features due to a broad user base.

I think this project is useful in case one wants to use attrs (pydantic only supports its own model and dataclasses) or something slim. However, I cannot spare the time for developing here, sorry.
